src/vanilla/utils/proxyMap.ts

```diff
@@ -1,111 +1,166 @@
+const maybeProxify = (x: any) => (typeof x === 'object' ? proxy({ x }).x : x)
+const isProxy = (x: any) => proxyStateMap.has(x)
 
-type InternalProxyMap<K, V> = Map<K, V> & {
-  data: KeyValRecord<K, V>[]
-  toJSON: object
+type InternalProxyObject<K, V> = Map<K, V> & {
+  data: Array<K | V>
+  index: number
+  toJSON: () => Map<K, V>
 }
 
-/**
- * proxyMap
- *
- * This is to create a proxy which mimic the native Map behavior.
- * The API is the same as Map API
- *
- * @example
- * import { proxyMap } from 'valtio/utils'
- * const state = proxyMap([["key", "value"]])
- *
- * //can be used inside a proxy as well
- * const state = proxy({
- *   count: 1,
- *   map: proxyMap()
- * })
- *
- * // When using an object as a key, you can wrap it with `ref` so it's not proxied
- * // this is useful if you want to preserve the key equality
- * import { ref } from 'valtio'
- *
- * const key = ref({})
- * state.set(key, "value")
- * state.get(key) //value
- *
- * const key = {}
- * state.set(key, "value")
- * state.get(key) //undefined
- */
-export function proxyMap<K, V>(entries?: Iterable<readonly [K, V]> | null) {
-  const map: InternalProxyMap<K, V> = proxy({
-    data: Array.from(entries || []) as KeyValRecord<K, V>[],
-    has(key) {
-      return this.data.some((p) => p[0] === key)
+export function proxyMap<K, V>(entries?: Iterable<[K, V]> | undefined | null) {
+  const initialData: Array<K | V> = []
+  let initialIndex = 0
+  const indexMap = new Map<K, number>()
+
+  const snapMapCache = new WeakMap<object, Map<K, number>>()
+  const registerSnapMap = () => {
+    const cache = snapCache.get(vObject)
+    const latestSnap = cache?.[1]
+    if (latestSnap && !snapMapCache.has(latestSnap)) {
+      const clonedMap = new Map(indexMap)
+      snapMapCache.set(latestSnap, clonedMap)
+      return true
+    }
+    return false
+  }
+  const getSnapMap = (x: any) => snapMapCache.get(x)
+
+  if (entries) {
+    if (typeof entries[Symbol.iterator] !== 'function') {
+      throw new TypeError(
+        'proxyMap:\n\tinitial state must be iterable\n\t\ttip: structure should be [[key, value]]',
+      )
+    }
+    for (const [k, v] of entries) {
+      const key = maybeProxify(k)
+      const value = maybeProxify(v)
+      indexMap.set(key, initialIndex)
+      initialData[initialIndex++] = key
+      initialData[initialIndex++] = value
+    }
+  }
+
+  const vObject: InternalProxyObject<K, V> = {
+    data: initialData,
+    index: initialIndex,
+    get size() {
+      if (!isProxy(this)) {
+        registerSnapMap()
+      }
+      const map = getSnapMap(this) || indexMap
+      return map.size
+    },
+    get(key: K) {
+      const map = getSnapMap(this) || indexMap
+      const k = maybeProxify(key)
+      const index = map.get(k)
+      if (index === undefined) {
+        if (!isProxy(this)) {
+          // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+          this.index
+        }
```

Comment:
We could make it `this.index` only, which should be more performant, because if it's a proxy, it's almost nothing, and probably faster than `isProxy()`.